quizzes/ch04-02-references-and-snapshots.toml

```
@@ -0,0 +1,140 @@
+[[questions]]
+type = "Tracing"
+prompt.program = """
+fn incr(ref n: u32) {
+    *n += 1;
+}
+
+fn main() {
+    let mut n : u32 = 1;
+    incr(ref n);
+    println!("{}", n);
+}
+"""
+answer.doesCompile = false
+context = """
+The code does not compile because the *desnap* operator has to be used on snapshots, not references.
+Inside the `incr` function, the `n` parameter is a reference to a `u32` value, so the `*n` operation is not allowed. The code should be changed to `n += 1;`.
+"""
+[[questions]]
+type = "Tracing"
+prompt.program = """
+#[derive(Drop)]
+struct Rectangle {
+    height: u64,
+    width: u64,
+}
+
+fn double_rect(ref rect: Rectangle) {
+    rect.height *= 2;
+    rect.width *= 2;
+}
+
+fn main() {
+    let rect = Rectangle { height: 10, width: 20 };
+    double_rect(ref rect);
+    println!("{}", rect.height);
+}
+"""
+answer.doesCompile = false
+context = """
+The code does not compile because a `ref` can only be used on mutable variables.
+You should write `let mut rect = Rectangle { height: 10, width: 20 };` instead.
+"""
+[[questions]]
+type = "Tracing"
+prompt.program = """
+fn calculate_length(arr: @Array<u128>) -> usize {
```

1. calculate_length is just doing `arr.len()` perhaps something like `sum` that sums all array elements using a loop would be nice?

```suggestion
fn sum(arr: @Array<u128>) -> u128 {
  let mut span = arr.span();
  let mut sum = 0;
  while let Option::Some(x) = span.pop_front() {
    sum += *x;
  };
  sum
}

```